lib/backend/etcdv3/etcdv3.go

```diff
@@ -0,0 +1,462 @@
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"github.com/coreos/etcd/pkg/transport"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	"github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+)
+
+var (
+	clientTimeout = 30 * time.Second
+)
+
+type EtcdV3Client struct {
+	etcdClient *clientv3.Client
+}
+
+func NewEtcdV3Client(config *apiconfig.EtcdConfig) (api.Client, error) {
+	// Determine the location from the authority or the endpoints.  The endpoints
+	// takes precedence if both are specified.
+	etcdLocation := []string{}
+	if config.EtcdEndpoints != "" {
+		etcdLocation = strings.Split(config.EtcdEndpoints, ",")
+	}
+
+	if len(etcdLocation) == 0 {
+		log.Info("No etcd endpoints specified in etcdv3 API config")
+		return nil, goerrors.New("no etcd endpoints specified")
+	}
+
+	// Create the etcd client
+	tlsInfo := &transport.TLSInfo{
+		CAFile:   config.EtcdCACertFile,
+		CertFile: config.EtcdCertFile,
+		KeyFile:  config.EtcdKeyFile,
+	}
+	tls, _ := tlsInfo.ClientConfig()
+
+	cfg := clientv3.Config{
+		Endpoints:   etcdLocation,
+		TLS:         tls,
+		DialTimeout: clientTimeout,
+	}
+
+	// Plumb through the username and password if both are configured.
+	if config.EtcdUsername != "" && config.EtcdPassword != "" {
+		cfg.Username = config.EtcdUsername
+		cfg.Password = config.EtcdPassword
+	}
+
+	client, err := clientv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EtcdV3Client{etcdClient: client}, nil
+}
+
+// Create an entry in the datastore.  If the entry already exists, this will return
+// an ErrorResourceAlreadyExists error and the current entry.
+func (c *EtcdV3Client) Create(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"model-key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	logCxt.Debug("Processing Create request")
+	key, value, err := getKeyValueStrings(d)
+	if err != nil {
+		return nil, err
+	}
+	logCxt = logCxt.WithField("etcdv3-key", key)
+
+	putOpts, err := c.getTTLOption(d)
+	if err != nil {
+		return nil, err
+	}
+
+	// Checking for 0 version of the key, which means it doesn't exists yet,
+	// and if it does, get the current value.
+	logCxt.Debug("Performing etcdv3 transaction for Create request")
+	txnResp, err := c.etcdClient.Txn(context.Background()).If(
+		clientv3.Compare(clientv3.Version(key), "=", 0),
+	).Then(
+		clientv3.OpPut(key, value, putOpts...),
+	).Else(
+		clientv3.OpGet(key),
+	).Commit()
+	if err != nil {
+		logCxt.WithError(err).Warning("Create failed")
+		return nil, errors.ErrorDatastoreError{Err: err}
+	}
+
+	if !txnResp.Succeeded {
+		// The resource must already exist.  Extract the current value and
+		// return that if possible.
+		logCxt.Info("Create transaction failed due to resource already existing")
```

Comment:
is this always the case when `txnResp.Scceeded` = `false`?

Reply:
I believe the Succeeded is a response to the If() portion of the transaction, and so false would be because the resource already exists.  However, since you are the primary reviewer for this it might be good to dive into the etcdv3 calls to check my assumptions.

Reply:
There are these server error codes the etcdv3 client can return https://github.com/coreos/etcd/blob/master/etcdserver/api/v3rpc/rpctypes/error.go#L23

See example in https://godoc.org/github.com/coreos/etcd/clientv3
```go
resp, err := kvc.Put(ctx, "", "")
if err != nil {
	if err == context.Canceled {
		// ctx is canceled by another routine
	} else if err == context.DeadlineExceeded {
		// ctx is attached with a deadline and it exceeded
	} else if verr, ok := err.(*v3rpc.ErrEmptyKey); ok {
		// process (verr.Errors)
	} else {
		// bad cluster endpoints, which are not etcd servers
	}
}
```